### PR TITLE
Pattern supports

### PR DESCRIPTION
**Pull request by coditive-adam-s**
This PR adds pattern support for the plugin. It allows the creation of patterns by providing the HTML markup inside the "resources/patterns" directory.

### CODE REVIEW

**Review by coditive-pawel-m — CHANGES_REQUESTED** [review]
@coditive-adam-s pls, take into account scenario when there's no tag (or content in tag) `<title>` or `<description>`. In such case we shouldn't register pattern.

File: src/Blocks/Blocks.php [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 namespace AstraticBlocks\Blocks;
 
+use AstraticBlocks\Blocks\Base;
```
Comment (coditive-michal-s) [inline src/Blocks/Blocks.php:5]: why is this here?
looks like the class is not used here 🤔

File: src/Patterns/Parser.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        $openTag = "<categories (.+?)>";
+        $closeTag = "</categories>";
+          break;
+      default:
+        $openTag = "<astratic-pattern>";
+        $closeTag = "</astratic-pattern>";
+          break;
+    };
+
+    preg_match("~$openTag(.+)$closeTag~misU", $content, $matches);
```
Comment (coditive-michal-s) [inline src/Patterns/Parser.php:28]: While it works the tilde `~` operator is not that common in regex imho so I'd change it to the `/`.
Also, the open and close tags are basically the same so they could be combined into the pattern right away.

What would you say about sth like this:
```
switch ($metaType) {
      case 'title':
            $pattern = "\<title\>(.+)\<\/title\>";
          break;
      // (...)
};

preg_match("/{$pattern}/misU", $content, $matches);
// ...
```

File: src/Patterns/Patterns.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+use WP_Block_Patterns_Registry;
+use WP_Block_Pattern_Categories_Registry;
+
+class Patterns
+{
+  private string $dir;
+  private array $patterns;
+  private array $allowedFiles;
+  private Parser $parser;
+  private const ABSL_PATTERN_MAIN_CATEGORY = 'Astratic Patterns';
```
Comment (coditive-michal-s) [inline src/Patterns/Patterns.php:15]: I believe it should be A **_SB_** L but this is a "class" const anyway so no need for prefixing imho 😅

File: src/Patterns/Patterns.php [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+  public function __construct()
+  {
+    $this->dir = ASBL_PATTERNS_PATH;
+    $this->patterns = array_filter(scandir($this->dir), function ($item) {
```
Comment (coditive-michal-s) [inline src/Patterns/Patterns.php:20]: what about [glob](https://www.php.net/manual/en/function.glob.php)? did you check it?

File: src/Blocks/Blocks.php (hunk 1 above)
Comment (coditive-adam-s) [reply src/Blocks/Blocks.php:5]: This is a small improvement for the code from previous work here: https://github.com/wpserved/astratic-blocks/pull/4/files#diff-16373bb60f50bc0484b2beb32047fc2945e0493d2cc407d556f0015129aefff0. I've talked with @coditive-przemek-h about it and he said there is no need to create new branch only to add this one line.

File: src/Patterns/Patterns.php (hunk 4 above)
Comment (coditive-adam-s) [reply src/Patterns/Patterns.php:20]: Yes, I've checked glob(). However doesn't found any advantages over the code I've have wrote with scandir().

File: src/Patterns/Patterns.php [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+  private array $allowedFiles;
+  private Parser $parser;
+  private const PATTERN_MAIN_CATEGORY = 'Astratic Patterns';
+
+  public function __construct()
+  {
+    $this->dir = ASBL_PATTERNS_PATH;
+    $this->patterns = array_filter(scandir($this->dir), function ($item) {
+      return ! is_dir($this->dir . '/' . $item) && array_reverse(explode('.', $item))[0] == 'html';
+    });
```
Comment (coditive-przemek-h) [inline src/Patterns/Patterns.php:22]: New line after this?

**Review by coditive-przemek-h — APPROVED** [review]

**Review by coditive-michal-s — APPROVED** [review]

**Review by coditive-pawel-m — APPROVED** [review]
